File: Python/Easy/217_contains_duplicate.py
"""
217. Contains Duplicate

Given an integer array nums, return true if any value appears at least twice
in the array, and return false if every element is distinct.

Constraints:
- 1 <= nums.length <= 10^5
- -10^9 <= nums[i] <= 10^9

Examples:

Example 1:
    Input: nums = [1,2,3,1]
    Output: true

Example 2:
    Input: nums = [1,2,3,4]
    Output: false

Example 3:
    Input: nums = [1,1,1,3,3,4,3,2,4,2]
    Output: true
"""


class Solution:
    def containsDuplicate(self, nums: list[int]) -> bool:
        """
        Returns True if any value appears at least twice.

        Converts nums to a set, discarding duplicates. If the set is smaller
        than the original list, at least one duplicate exists. Always processes
        the entire list.

        Time complexity:  O(n)
        Space complexity: O(n) for the set
        """
        return len(set(nums)) != len(nums)

    def containsDuplicateEarlyExit(self, nums: list[int]) -> bool:
        """
        Returns True if any value appears at least twice, with early exit.

        Iterates once, inserting each element into a seen set. Returns True
        immediately on the first repeated element, avoiding unnecessary work
        when a duplicate appears early in a large list.

        Time complexity:  O(n) worst case; faster in practice with early exit
        Space complexity: O(n) for the seen set
        """
        seen: set[int] = set()
        for num in nums:
            if num in seen:
                return True
            seen.add(num)
        return False


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ([1, 2, 3, 1], True),
        ([1, 2, 3, 4], False),
        ([1, 1, 1, 3, 3, 4, 3, 2, 4, 2], True),
        ([1], False),
    ]

    for label, method in [
        ("containsDuplicate         ", solution.containsDuplicate),
        ("containsDuplicateEarlyExit", solution.containsDuplicateEarlyExit),
    ]:
        for nums, expected in test_cases:
            result = method(nums)
            assert result == expected, \
                f"{label} failed: nums={nums} → expected {expected}, got {result}"
        print(f"  {label} passed")

    print("All test cases passed!")

File: Golang/Easy/217_contains_duplicate.go
package main

import "fmt"

// containsDuplicate reports whether any value in nums appears at least twice.
//
// Inserts every element into a map[int]struct{} (an idiomatic Go set), then
// compares the map length to the slice length. Always processes the entire
// slice even if a duplicate is found early.
//
// Time complexity:  O(n)
// Space complexity: O(n)
func containsDuplicate(nums []int) bool {
	set := make(map[int]struct{}, len(nums))
	for _, item := range nums {
		set[item] = struct{}{}
	}
	return len(nums) != len(set)
}

// containsDuplicateEarlyExit reports whether any value in nums appears at
// least twice, returning true as soon as the first duplicate is found.
//
// Time complexity:  O(n) worst case; faster in practice with early exit
// Space complexity: O(n)
func containsDuplicateEarlyExit(nums []int) bool {
	seen := make(map[int]bool)
	for _, n := range nums {
		if seen[n] {
			return true
		}
		seen[n] = true
	}
	return false
}

func main() {
	testCases := []struct {
		nums     []int
		expected bool
	}{
		{[]int{1, 2, 3, 1}, true},
		{[]int{1, 2, 3, 4}, false},
		{[]int{1, 1, 1, 3, 3, 4, 3, 2, 4, 2}, true},
		{[]int{1}, false},
	}

	for i, tc := range testCases {
		r1, r2 := containsDuplicate(tc.nums), containsDuplicateEarlyExit(tc.nums)
		s1, s2 := "PASS", "PASS"
		if r1 != tc.expected {
			s1 = "FAIL"
		}
		if r2 != tc.expected {
			s2 = "FAIL"
		}
		fmt.Printf("Example %d | Expected: %v | set: %v [%s] | earlyExit: %v [%s]\n",
			i+1, tc.expected, r1, s1, r2, s2)
	}
}
